README.md
# egg-websocket-plugin

[![NPM version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]
[![Test coverage][codecov-image]][codecov-url]
[![David deps][david-image]][david-url]
[![Known Vulnerabilities][snyk-image]][snyk-url]
[![npm download][download-image]][download-url]

[npm-image]: https://img.shields.io/npm/v/egg-websocket-plugin.svg?style=flat-square
[npm-url]: https://npmjs.org/package/egg-websocket-plugin
[travis-image]: https://img.shields.io/travis/flarestart/egg-websocket-plugin.svg?style=flat-square
[travis-url]: https://travis-ci.org/flarestart/egg-websocket-plugin
[codecov-image]: https://img.shields.io/codecov/c/github/flarestart/egg-websocket-plugin.svg?style=flat-square
[codecov-url]: https://codecov.io/github/flarestart/egg-websocket-plugin?branch=master
[david-image]: https://img.shields.io/david/flarestart/egg-websocket-plugin.svg?style=flat-square
[david-url]: https://david-dm.org/flarestart/egg-websocket-plugin
[snyk-image]: https://snyk.io/test/npm/egg-websocket-plugin/badge.svg?style=flat-square
[snyk-url]: https://snyk.io/test/npm/egg-websocket-plugin
[download-image]: https://img.shields.io/npm/dm/egg-websocket-plugin.svg?style=flat-square
[download-url]: https://npmjs.org/package/egg-websocket-plugin

<!--
Description here.
-->

## 依赖说明

### 依赖的 egg 版本

✅ egg 2.x



## 特性支持

- [x] 兼容 egg 自身的 controller
- [x] 兼容 egg 路由中间件
- [x] 不需要像 socket.io 一样开启 `sticky` 模式
- [x] 支持消息广播订阅



## 使用方式

```bash
yarn add egg-websocket-plugin
```

### 1. 开启插件

```js
// config/plugin.js
exports.websocket = {
  enable: true,
  package: 'egg-websocket-plugin',
};
```

### 2. 配置 WebSocket 路由

```js
// app/router.js
app.ws.route('/ws', app.controller.home.hello);

// 可使用路由参数，使用方式同 egg 自身路由
app.ws.route('/foo/:id', app.controller.home.foo);
```

### 3. 配置 WebSocket 全局中间件【可选】

```js
// app/router.js

// 配置 WebSocket 全局中间件
app.ws.use((ctx, next) => {
  console.log('websocket open');
  await next();
  console.log('websocket closed');
});
```

### 4. 配置路由中间件【可选】

**路由会依次用到 app.use, app.ws.use, 以及 app.ws.router 中配置的中间件**

```js
// app/router.js
function middleware(ctx, next) {
  console.log('open', ctx.starttime);
  return next();
}
// 配置路由中间件
app.ws.route('/ws', middleware, app.controller.home.hello);
```

### 5. 禁用 App 全局中间件

> 插件默认会使用 app.use(...) 注册的中间件，如果你不想使用它们，或者这些插件与 websocket 有冲突，你可以在 `config.default.js` 中禁用它们

```js
config.websocket = {
  useAppMiddlewares: false,
};
```

### 6. 在控制器中使用 websocket

websocket 是一个 `ws` 插件的实例，可阅读 [ws](https://www.npmjs.com/package/ws) 插件的说明文档或 TypeScript 的定义

```js
// app/controller/home.js
import { Controller } from 'egg';

export default class HomeController extends Controller {
  async hello() {
    const { ctx, app } = this;
    if (!ctx.websocket) {
      throw new Error('this function can only be use in websocket router');
    }

    console.log('client connected');

    ctx.websocket
      .on('message', (msg) => {
        console.log('receive', msg);
      })
      .on('close', (code, reason) => {
        console.log('websocket closed', code, reason);
      });
  }
}

```



## 支持 Redis 消息发布订阅【可选】

### 开启消息广播

**注意**：`egg-websocket-plugin` 依赖于 `ioredis` 但可与其他插件如 `egg-redis` 共用一个 `ioredis` 依赖

开启广播相关的 API 需要在 config.default.js 中配置 `redis` 连接，配置项可参见 [ioredis](https://www.npmjs.com/package/ioredis)

```typescript
config.websocket = {
  // 配置 websocket 使用 redis 作消息广播，配置项目参见 ioredis
  redis: {
    host: '127.0.0.1',
    port: 6379,
  },
};
```

### 房间消息广播 API

> 开启 redis 配置后，即可使用下面的广播 API

#### 1. app.ws.sendTo(room, data)

> 发送消息到某个房间，将 data 原样发送到客户端连接

- `room` &lt;string&gt;: 房间名
- `data` &lt;string | Buffer&gt;: 发送内容，原样发送

#### 2. app.ws.sendJsonTo(room, data)

> 发送消息到某个房间，会自动使用 JSON.stringify 对 data 进行序列化

- `room` &lt;string&gt; 房间名
- `data` &lt;any&gt; 发送内容，可进行 JSON 序列化的任意对象

#### 3. ctx.websocket.room.join(room, fn?)
> 将当前连接加入房间

- `room` &lt;string&gt; 要加入的房间名
- `fn` &lt;function({room, message})&gt; 可选参数，用于配置房间消息处理函数，配置详情见 [自定义房间消息处理函数](#roomhandler)
  - `room` &lt;string&gt; 消息所属房间名
  - `message`<string | Buffer> 房间内的广播消息，消息类型取决于 `sendTo()` 发送时的类型
  - 忽略此参数，则该房间的消息将直接发送到客户端连接
  - 配置这个参数，则可以手动处理消息发送逻辑，不再自动将房间内的消息发送到客户端连接

#### 4. ctx.websocket.room.leave(room)
> 当前连接离开房间，当前连接关闭后，将自动退出当前连接加入过的房间，因此关闭连接前无需要手动离开房间

- `room` &lt;string&gt; 房间名

#### 5. ctx.websocket.room.sendTo(room, data)

> 发送消息到某个房间，将 data 原样发送到客户端连接，功能与 app.ws.sendTo 相同

#### 6. ctx.websocket.room.sendJsonTo(room, data)

> 发送消息到某个房间，会自动使用 JSON.stringify 对 data 进行序列化，功能与 `app.ws.sendTo` 相同



```typescript
// 当前连接加入房间 foo，收到房间的消息将直接发送到当前客户端连接
ctx.websocket.room.join('foo');

// 当前连接加入房间 bar，并设置该房间消息处理函数，设置了处理函数将不再直接发送消息到当前客户端
// room: string, message: Buffer | string
ctx.websocket.room.join('bar', ({ room, message }) => {
  console.log(room, message);
});

// 当前连接离开房间
ctx.websocket.room.leave('foo');

// 通过当前 ctx 广播消息，所有加入房间的连接会收到这条消息
ctx.websocket.room.sendTo('foo', 'hello foo');

// 也可以通过通过 app 对象来广播消息
app.ws.sendTo('bar', 'hello bar');
```





### 自定义房间消息处理函数<a id="roomhandler"></a>

**注意**：使用 `room.join()` 函数，忽略第二个参数的情况下，房间内的消息会直接发送到当前客户端连接。

如果你需要自定义这部分的处理逻辑，那么可配置 `room.join()` 的第二个参数。

```typescript
ctx.websocket.room.join('foo', ({ room, message }) => {
  // room: string 房间名
  // message: Buffer | string 消息内容
  // message 的类型取决于使用 sendTo 发送的是 string 还是 Buffer
  console.log(room, message);
});
```



### 例子

router.ts

```typescript
import { Application } from 'egg';

export default (app: Application) => {
  const { controller, router } = app;

  // 配置 http 路由
  router.get('/', controller.home.index);
  // 配置 websocket 路由，websocket 可与 http 使用同一个 url，不会导致冲突
  app.ws.route('/', controller.home.ws);
};
```

controller/home.ts

```typescript
import { Controller } from 'egg';

export default class HomeController extends Controller {
  public async index() {
    const { ctx, app } = this;
    // 向房间 foo 发送消息
    app.ws.sendTo('foo', 'hello from index');
    ctx.body = 'index';
  }

  public async ws() {
    const { websocket } = this.ctx;
    if (!websocket) {
      throw new Error('this function can only be use in websocket router');
    }

    // 加入到房间 foo
    websocket.room.join('foo');
  }
}
```

浏览器连接 WebSocket

```javascript
const ws = new WebSocket('ws://127.0.0.1:7001/');
ws.addEventListener('message', data => {
  console.log(data);
});
```

这时候，访问 http://localhost:7001/ WebSocket 客户端将会收到一条  `hello from index` 的字符串



## 使用场景

- 使用 egg 开发标准 WebSocket 服务器
- 微信小程序 WebSocket 服务器



## 提问交流

请到 [issues](https://github.com/flarestart/egg-websocket-plugin/issues) 异步交流。



## License

[MIT](LICENSE)

CHANGELOG.md
## <small>1.0.1 (2020-12-29)</small>

* refactor: 部分代码重构 ([6fca207](https://github.com/flarestart/egg-websocket-plugin/commit/6fca207))
* test: 添加单元测试 ([758213d](https://github.com/flarestart/egg-websocket-plugin/commit/758213d))



# [1.0.0](https://github.com/flarestart/egg-websocket-plugin/compare/v0.2.1...v1.0.0) (2020-08-22)


### feat

* 添加房间，允许接受房间广播、加入、离开 ([277ed18](https://github.com/flarestart/egg-websocket-plugin/commit/277ed185f017a1b4640a99772c6019a500bc3a37))

### fix

* WebSocket 被中间件拦截后或捕获到错误后，正确的关闭连接 ([026aef5](https://github.com/flarestart/egg-websocket-plugin/commit/026aef57894e92124a6b11bc1639364e72d2cda4))
* 避免在未配置 Redis 的情况下，使用广播方法可能导致未捕获的错误 ([b7b0ed7](https://github.com/flarestart/egg-websocket-plugin/commit/b7b0ed703685e3def8baf21783d2ab16febdf8df))



## [0.1.1](https://github.com/flarestart/egg-websocket-plugin/compare/v0.1.0...v0.1.1) (2020-07-18)


### chore

* 文档更新 ([628c680](https://github.com/flarestart/egg-websocket-plugin/commit/628c6806a4b8271ee0d1681d58e44e7adf904470))

### fix

* 使用 egg 自身的 Router 来处理路由 ([9e87019](https://github.com/flarestart/egg-websocket-plugin/commit/9e87019dc52a3a0a14e5a41ae3d43fa5607f4169))
* 修复默认配置错误 ([fd6fcfa](https://github.com/flarestart/egg-websocket-plugin/commit/fd6fcfa49ce96d2a2a5539dbdee71132f3cb273f))



## [0.0.3](https://github.com/flarestart/egg-websocket-plugin/compare/v0.0.2...v0.0.3) (2020-04-16)


### fix

* TypeScript 定义文件更新 ([355a82b](https://github.com/flarestart/egg-websocket-plugin/commit/355a82b455b29895e52e8f3149446c8b768063ce))



## [0.0.2](https://github.com/flarestart/egg-websocket-plugin/compare/227a81c38ea6a147e40b04fc61e63ad346514fa8...v0.0.2) (2020-01-06)


### chore

* 基础代码 ([227a81c](https://github.com/flarestart/egg-websocket-plugin/commit/227a81c38ea6a147e40b04fc61e63ad346514fa8))

### fix

* 中间件优化；文档更新 ([36ebf6b](https://github.com/flarestart/egg-websocket-plugin/commit/36ebf6b2036cb3958d57c0af0039a4fb861d99c6))





test/room.test.ts
'use strict';

import * as http from 'http';
import { MockApplication } from 'egg-mock';
import * as WebSocket from 'ws';
import { EggWsServer } from '../app';

// const assert = require('assert');
const mock = require('egg-mock');
const mockServer = require('egg-mock/lib/mock_http_server');

interface MockApp extends MockApplication {
  ws: EggWsServer;
}

let wsUrl: string;
function WebSocketClient(url: string) {
  return new WebSocket(`${wsUrl}${url}`);
}

describe('test/websocket-plugin.test.js', () => {
  let app: MockApp;
  let server: http.Server;
  before(async () => {
    app = mock.app({
      baseDir: 'apps/websocket-plugin-test',
    });
    await app.ready();
    server = mockServer(app);
    server.listen();
    const addr = server.address();
    let port = '';
    if (typeof addr === 'string') {
      throw new Error('unexpected server address');
    } else if (addr === null) {
      throw new Error('websocket server listen failed');
    } else {
      port = `${addr.port}`;
    }
    wsUrl = `ws://127.0.0.1:${port}`;
  });

  after(() => app.close());
  afterEach(mock.restore);

  it('websocket error', async () => {
    WebSocketClient('/error');
  });

  it('websocket controller error', async () => {
    app.ws.use(async (_, next) => {
      try {
        await next();
      } catch (e) {}
    });
    WebSocketClient('/controller/error');
  });

  it('websocket controller error', async () => {
    app.ws.use(async (ctx, next) => {
      console.log(ctx.headers);
      await next();
    });
  });

  it('join room', async () => {
    app.ws.route('/room/join', ctx => {
      ctx.websocket.room.join(['room1', 'room2'], ({ room, message }) => {
        console.log(room, message);
      });
    });

    const ws = WebSocketClient('/room/join');
    await new Promise(resolve => {
      ws.on('open', () => {
        resolve();
      });
    });
    ws.close();
  });
});


test/no-adapter.test.ts
'use strict';

import * as http from 'http';
import { MockApplication } from 'egg-mock';
import * as WebSocket from 'ws';
import { EggWsServer } from '../app';

const assert = require('assert');
const mock = require('egg-mock');
const mockServer = require('egg-mock/lib/mock_http_server');

interface MockApp extends MockApplication {
  ws: EggWsServer;
}

describe('test/no-adapter.test.ts', () => {
  let app: MockApp;
  let server: http.Server;
  let wsUrl: string;
  function NewTestWebSocket(url: string) {
    return new WebSocket(`${wsUrl}${url}`);
  }

  before(async () => {
    app = mock.app({
      baseDir: 'apps/websocket-plugin-no-adapter-test',
    });
    await app.ready();
    server = mockServer(app);
    server.listen();
    const addr = server.address();
    let port = '';
    if (typeof addr === 'string') {
      throw new Error('unexpected server address');
    } else if (addr === null) {
      throw new Error('websocket server listen failed');
    } else {
      port = `${addr.port}`;
    }
    wsUrl = `ws://127.0.0.1:${port}`;
  });

  after(() => app.close());
  afterEach(mock.restore);

  it('should not use room.join without adapter', async () => {
    const url = '/no-adapter/room/join';
    app.ws.route(url, ctx => {
      ctx.websocket.room.join('room1');
    });
    await new Promise<void>(resolve => {
      app.ws.server.once('error', e => {
        assert(e.message.includes('adapter configure not found'));
        resolve();
      });
      const client = NewTestWebSocket(url);
      client.on('open', () => {
        client.close();
      });
    });
  });

  it('should not use room.leave without adapter', async () => {
    const url = '/no-adapter/room/leave';
    app.ws.route(url, ctx => {
      ctx.websocket.room.leave('room1');
    });
    await new Promise<void>(resolve => {
      app.ws.server.once('error', e => {
        assert(e.message.includes('adapter configure not found'));
        resolve();
      });
      const client = NewTestWebSocket(url);
      client.on('open', () => {
        client.close();
      });
    });
  });

  it('should not use app.sendTo without adapter', async () => {
    await new Promise<void>(resolve => {
      app.ws.server.once('error', e => {
        assert(e.message.includes('adapter configure not found'));
        resolve();
      });
      app.ws.sendTo('room1', 'hello');
    });
  });

  it('should not use app.sendJsonTo without adapter', async () => {
    await new Promise<void>(resolve => {
      app.ws.server.once('error', e => {
        assert(e.message.includes('adapter configure not found'));
        resolve();
      });
      app.ws.sendJsonTo('room1', 'hello');
    });
  });

  it('should not use ctx.room.sendTo without adapter', async () => {
    await new Promise<void>(resolve => {
      app.ws.server.once('error', e => {
        assert(e.message.includes('adapter configure not found'));
        resolve();
      });
      app.ws.sendJsonTo('room1', 'hello');
    });
  });

  it('should not use ctx.room.sendTo without adapter', async () => {
    const url = '/no-adapter/room/sendTo';
    app.ws.route(url, ctx => {
      ctx.websocket.room.sendTo('room1', 'hello');
    });
    await new Promise<void>(resolve => {
      app.ws.server.once('error', e => {
        assert(e.message.includes('adapter configure not found'));
        resolve();
      });
      const client = NewTestWebSocket(url);
      client.on('open', () => {
        client.close();
      });
    });
  });


  it('should not use ctx.room.sendJsonTo without adapter', async () => {
    const url = '/no-adapter/room/sendJsonTo';
    app.ws.route(url, ctx => {
      ctx.websocket.room.sendJsonTo('room1', 'hello');
    });
    await new Promise<void>(resolve => {
      app.ws.server.once('error', e => {
        assert(e.message.includes('adapter configure not found'));
        resolve();
      });
      const client = NewTestWebSocket(url);
      client.on('open', () => {
        client.close();
      });
    });
  });
});


test/websocket-plugin.test.js
'use strict';

const assert = require('assert');
const mock = require('egg-mock');
const ws = require('ws');

describe('test/websocket-plugin.test.js', () => {
  let app;
  before(() => {
    app = mock.app({
      baseDir: 'apps/websocket-plugin-test',
    });
    return app.ready();
  });

  after(() => app.close());
  afterEach(mock.restore);

  it('should have app ws property', () => {
    assert(app.ws instanceof ws.Server);
  });

  it('should have route function', () => {
    assert(typeof app.ws.route === 'function');
  });

  it('should have 404 from http request ', () => {
    app.ws.route('/ws', ctx => {
      ctx.websocket.close();
    });
    return app
      .httpRequest()
      .get('/ws')
      .expect(404);
  });
});


app.js
'use strict';
const assert = require('assert');
const ws = require('ws');
const compose = require('koa-compose');
const url = require('url');
const http = require('http');
const Router = require('@eggjs/router');

function isFunction(fn) {
  return typeof fn === 'function';
}

function isString(str) {
  return typeof str === 'string';
}

// Egg Websocket Plugin
class EggWebsocket extends ws.Server {
  static resolveController(controller, app) {
    if (isString(controller)) {
      const actions = controller.split('.');
      let obj = app.controller;
      actions.forEach(key => {
        obj = obj[key];
        assert(
          isFunction(obj),
          `[egg-websocket-plugin]: controller '${controller}' not exists`
        );
      });
      controller = obj;
    }
    // ensure controller is exists
    assert(
      isFunction(controller),
      '[egg-websocket-plugin]: controller not exists'
    );
    return controller;
  }

  constructor(app) {
    super({
      noServer: true,
    });
    this.app = app;
    this.router = new Router();
    this.middlewares = [];
    this.on('error', e => {
      app.logger.error('[egg-websocket-plugin] error: ', e);
    });
  }

  use(middleware) {
    assert(
      isFunction(middleware),
      '[egg-websocket-plugin] middleware should be a function'
    );
    if (this.middlewares.includes(middleware)) {
      return;
    }
    this.middlewares.push(middleware);
  }

  route(path, ...middleware) {
    assert(middleware.length > 0, '[egg-websocket-plugin] controller not set');
    // get last middleware as handler
    const handler = middleware.pop();
    let controller;
    if (isString(handler)) {
      // resolve handler from app's controller
      controller = EggWebsocket.resolveController(handler, this.app);
    } else {
      controller = handler;
    }
    const app = this.app;

    // check if need to use app middlewares
    let appMiddlewares = [];
    if (
      app.config.websocket &&
      app.config.websocket.useAppMiddlewares === true
    ) {
      appMiddlewares = app.middleware;
    }

    const composedMiddleware = compose([
      ...appMiddlewares,
      ...this.middlewares,
      ...middleware,
      waitWebSocket(controller),
    ]);

    this.router.all(path, composedMiddleware);
  }

  notFound(socket) {
    socket.end('HTTP/1.1 404 Not Found\r\n\r\n');
  }
}

// 运行 controller 并等待退出
function waitWebSocket(controller) {
  return ctx => {
    return new Promise((resolve, reject) => {
      ctx.websocket.on('close', resolve);
      ctx.websocket.on('error', reject);
      try {
        const ret = controller.call(ctx);
        if (ret instanceof Promise) {
          ret.catch(reject);
        }
      } catch (e) {
        reject(e);
      }
    });
  };
}

module.exports = app => {
  const ws = new EggWebsocket(app);
  // add ws object to app
  app.ws = ws;

  const upgradeHandler = (request, socket, head) => {
    if (!request.url) {
      return ws.notFound(socket);
    }

    const pathname = url.parse(request.url).pathname;
    const matches = ws.router.match(pathname, 'GET');

    // check if the route has a handler or not
    if (!matches.route) {
      return ws.notFound(socket);
    }
    const controller = ws.router.routes();
    // upgrade to websocket connection
    ws.handleUpgrade(request, socket, head, conn => {
      ws.emit('connection', conn, request);
      const ctx = app.createContext(request, new http.ServerResponse(request));
      ctx.websocket = conn;
      controller(ctx).catch(e => {
        // close websocket connection
        if (!ctx.websocket.CLOSED) {
          ctx.websocket.close();
        }
        ctx.onerror(e);
      });
    });
  };
  // add websocket upgrade event listener
  app.on('server', server => {
    server.on('upgrade', upgradeHandler);
  });
};


adapter/redis.ts
import { PubSuber } from '../app';
import { Redis, RedisOptions } from 'ioredis';

export class RedisPubSuber implements PubSuber {
  static TYPE_BINARY = 'b';
  static TYPE_STRING = 's';

  publisher: Redis;
  subscriber: Redis;
  joinedRoomCounter: Map<string, number> = new Map();
  handlers = new Set<Function>();
  listening = false;

  constructor(config: RedisOptions) {
    const IORedis = require('ioredis');
    this.publisher = new IORedis(config);
    this.subscriber = new IORedis(config);
  }

  publish(room: string, message: any) {
    if (typeof message === 'string') {
      return this.publisher.publish(
        room,
        `${RedisPubSuber.TYPE_STRING}${message}`
      );
    }
    return this.publisher.publish(
      room,
      `${RedisPubSuber.TYPE_BINARY}${message.toString()}`
    );
  }

  addMessageHandler(listener: Function) {
    this.handlers.add(listener);
    if (!this.listening) {
      this.listening = true;
      this.subscriber.addListener('messageBuffer', this.handleMessage);
    }
  }

  removeMessageHandler(listener: Function) {
    this.handlers.delete(listener);
    if (this.handlers.size <= 0 && this.listening) {
      this.listening = false;
      this.subscriber.removeListener('messageBuffer', this.handleMessage);
    }
  }

  handleMessage = (room: string | ArrayBufferLike, msg: any) => {
    const m = this._processMessage(msg);
    if (!m) {
      return;
    }
    const { message } = m;
    this.handlers.forEach(item => {
      item({ room: room.toString(), message });
    });
  };

  _processMessage(message: Buffer) {
    const messageType = String.fromCharCode(message[0]);
    if (messageType === RedisPubSuber.TYPE_BINARY) {
      return {
        type: 'binary',
        message: message.slice(1),
      };
    }
    if (messageType === RedisPubSuber.TYPE_STRING) {
      return {
        type: 'string',
        message: message.slice(1).toString(),
      };
    }
    return null;
  }

  joinRoom(...rooms: string[]) {
    const newRooms: string[] = [];
    rooms.forEach(room => {
      const isNew = this._increaseRoomCounter(room);
      if (isNew) {
        newRooms.push(room);
      }
    });
    if (newRooms.length <= 0) {
      return Promise.resolve(0);
    }
    return this.subscriber.subscribe(...newRooms);
  }

  leaveRoom(...rooms: string[]) {
    const leaveRooms: any[] = [];
    rooms.forEach(room => {
      const isDelete = this._decreaseRoomCounter(room);
      if (isDelete) {
        leaveRooms.push(room);
      }
    });
    if (leaveRooms.length <= 0) {
      return Promise.resolve(0);
    }
    return this.subscriber.unsubscribe(...leaveRooms);
  }

  // 房间计数器，计数器 <= 0 则取消订阅房间, 返回值 true: 新房间, false: 房间已订阅
  private _increaseRoomCounter(room: string): boolean {
    const count = this.joinedRoomCounter.get(room);
    if (!count || count <= 0) {
      this.joinedRoomCounter.set(room, 1);
      return true;
    }
    this.joinedRoomCounter.set(room, count + 1);
    return false;
  }
  private _decreaseRoomCounter(room: string): boolean {
    const count = this.joinedRoomCounter.get(room);
    if (!count || count <= 1) {
      this.joinedRoomCounter.delete(room);
      return true;
    }
    this.joinedRoomCounter.set(room, count - 1);
    return false;
  }
}


app.ts
import * as assert from 'assert';
import { Application, Context, EggAppConfig } from 'egg';
import { IncomingMessage, ServerResponse } from 'http';
import * as compose from 'koa-compose';
import { Socket } from 'net';
import * as url from 'url';
import * as WebSocket from 'ws';
import { EggRouter as Router } from '@eggjs/router';
import { RedisPubSuber } from './adapter/redis';

/**
 * RFC6455 defined status code
 * https://tools.ietf.org/html/rfc6455#section-7.4.1
 */
const WebSocketInternalError = 1011;

export interface EggWsClient extends WebSocket {
  room: EggWebsocketRoom;
}

export interface PubSuber {
  publish(room: string, message: any): void;
  addMessageHandler(listener: Function): void;
  removeMessageHandler(listener: Function): void;
  joinRoom(...rooms: string[]): Promise<number>;
  leaveRoom(...rooms: string[]): Promise<number>;
}

export type RoomHandler = (params: {
  room: string;
  message: ArrayBufferLike | string;
}) => void;

function isFunction(v: any): v is Function {
  return typeof v === 'function';
}

function isString(v: any): v is string {
  return typeof v === 'string';
}

// run controller until finish
function waitWebSocket(controller) {
  return ctx => {
    return new Promise((resolve, reject) => {
      ctx.websocket.on('close', resolve).on('error', reject);
      try {
        const ret = controller.call(ctx, ctx);
        if (ret && isFunction(ret.catch)) {
          ret.catch(reject);
        }
      } catch (err) {
        reject(err);
      }
    });
  };
}

export class EggWebsocketRoom {
  private _server: EggWsServer;
  private _ctx: Context;
  private _joinedRooms: Set<string>;
  private _listening: boolean;
  private _roomHandlers: Map<string, RoomHandler> = new Map();

  constructor(wsServer: EggWsServer, ctx: any) {
    this._server = wsServer;
    this._ctx = ctx;
    this._joinedRooms = new Set();
    this._listening = false;

    this._ctx.websocket!.on('close', () => {
      this.leave(Array.from(this._joinedRooms));
    });
  }

  private get adapter(): PubSuber | undefined {
    return this._server.adapter;
  }

  sendTo(room: string, data: ArrayBufferLike | string) {
    return this.adapter?.publish(room, data);
  }

  sendJsonTo(room: string, data: any) {
    return this.adapter?.publish(room, JSON.stringify(data));
  }

  join(rooms: string | string[], fn?: RoomHandler) {
    const adapter = this.adapter;
    if (!adapter) {
      return;
    }
    const newRooms: string[] = [];
    let handler: RoomHandler;
    if (!fn) {
      handler = this._defaultHandler;
    } else {
      handler = fn;
    }
    let roomsArray: string[];
    if (!Array.isArray(rooms)) {
      roomsArray = [rooms];
    } else {
      roomsArray = rooms;
    }
    roomsArray.forEach(room => {
      if (!this._joinedRooms.has(room)) {
        newRooms.push(room);
        this._joinedRooms.add(room);
      }
      this._roomHandlers.set(room, handler);
    });
    if (this._joinedRooms.size > 0 && !this._listening) {
      this._listening = true;
      adapter.addMessageHandler(this.onMessage);
    }
    return adapter.joinRoom(...newRooms);
  }

  leave(rooms: string | string[]) {
    if (rooms === '' || (Array.isArray(rooms) && rooms.length <= 0)) {
      return;
    }
    const adapter = this.adapter;
    if (!adapter) {
      return;
    }
    let roomsArray: string[];
    if (Array.isArray(rooms)) {
      roomsArray = rooms;
    } else if (isString(rooms)) {
      roomsArray = [rooms];
    } else {
      this._server.server.emit('error', new Error('invalid room name'));
      return;
    }
    const leaveRooms: string[] = [];
    roomsArray.forEach(room => {
      if (this._joinedRooms.has(room)) {
        leaveRooms.push(room);
        this._joinedRooms.delete(room);
      }
      this._roomHandlers.delete(room);
    });
    if (this._joinedRooms.size <= 0 && this._listening) {
      this._listening = false;
      adapter.removeMessageHandler(this.onMessage);
    }
    return adapter.leaveRoom(...leaveRooms);
  }

  private onMessage = ({
    room,
    message,
  }: {
    room: string;
    message: Buffer;
  }) => {
    const roomName = room.toString();
    const handlers = this._roomHandlers.get(roomName);
    if (handlers) {
      handlers({ room, message });
    }
  };

  private _defaultHandler = ({ message }) => {
    this._ctx.websocket!.send(message);
  };
}

// Egg Websocket Plugin
export class EggWsServer {
  static resolveController(controller, app) {
    if (isString(controller)) {
      const actions = controller.split('.');
      let obj = app.controller;
      actions.forEach(key => {
        obj = obj[key];
      });
      assert(
        isFunction(obj),
        `[egg-websocket-plugin]: controller '${controller}' not exists`
      );
      controller = obj;
    }
    // ensure controller is exists
    assert(
      isFunction(controller),
      '[egg-websocket-plugin]: controller not exists'
    );
    return controller;
  }

  server: WebSocket.Server;
  private _app: Application;
  private _adapter?: PubSuber;
  private _router = new Router();
  private _middlewares: any[];
  private _routerUsed = false;
  public clients = new Set();

  constructor(config: EggAppConfig['websocket'], app: Application) {
    this.server = new WebSocket.Server({
      noServer: true,
    });
    this._app = app;
    this._middlewares = [];
    this.server.on('error', e => {
      if (process.env['NODE_ENV'] === 'test') {
        return;
      }
      // istanbul ignore next
      app.logger.error('[egg-websocket-plugin] error: ', e.message);
    });

    if (config && config.redis) {
      this._adapter = new RedisPubSuber(config.redis);
    }

    app.on('server', server => {
      server.on('upgrade', this._upgradeHandler);
    });
  }

  private _upgradeHandler = (
    request: IncomingMessage,
    socket: Socket,
    head: Buffer
  ) => {
    // istanbul ignore next
    if (!request.url) {
      return this.notFound(socket);
    }

    const pathname = url.parse(request.url).pathname;
    const matches = this._router.match(pathname, 'GET');

    // check if the route has a handler or not
    if (!matches.route) {
      return this.notFound(socket);
    }
    const controller = this._router.routes();
    // upgrade to websocket connection
    this.server.handleUpgrade(request, socket, head, conn => {
      this.server.emit('connection', conn, request);

      const ctx = this._app.createContext(request, new ServerResponse(request));
      const expandConn: EggWsClient = conn as EggWsClient;
      ctx.websocket = expandConn;
      expandConn.room = new EggWebsocketRoom(this, ctx);
      this.clients.add(ctx);
      const closeHandler = (err?: Error) => {
        // close websocket connection
        this.clients.delete(ctx);
        if (ctx.websocket.readyState !== ctx.websocket.CLOSED) {
          ctx.websocket.close(err && WebSocketInternalError);
        }
      };
      controller(ctx)
        .then(() => {
          closeHandler();
        })
        .catch(e => {
          closeHandler(e);
          ctx.logger.error(e);
        });
    });
  };

  public get adapter(): PubSuber | undefined {
    if (!this._adapter) {
      const err = new Error(
        '[egg-websocket-plugin] redis pub/sub adapter configure not found'
      );
      this.server.emit('error', err);
    }
    return this._adapter;
  }

  use(middleware) {
    assert(
      isFunction(middleware),
      '[egg-websocket-plugin] middleware should be a function'
    );
    if (this._middlewares.includes(middleware)) {
      this._app.logger.warn(
        '[egg-websocket-plugin] same middleware has been used'
      );
      return;
    }
    if (this._routerUsed) {
      this._app.logger.warn(
        '[egg-websocket-plugin] app.ws.use should used before all app.ws.route'
      );
    }
    this._middlewares.push(middleware);
  }

  route(path, ...middleware) {
    assert(middleware.length > 0, '[egg-websocket-plugin] controller not set');
    // get last middleware as handler
    const handler = middleware.pop();
    const app = this._app;
    let controller;
    if (isString(handler)) {
      // resolve handler from app's controller
      controller = EggWsServer.resolveController(handler, app);
    } else {
      controller = handler;
    }
    this._routerUsed = true;

    // check if need to use app middlewares
    let appMiddlewares: any[] = [];
    if (
      app.config.websocket &&
      app.config.websocket.useAppMiddlewares === true
    ) {
      appMiddlewares = app.middleware;
    }

    const composedMiddleware = compose([
      ...appMiddlewares,
      ...this._middlewares,
      ...middleware,
      waitWebSocket(controller),
    ]);

    this._router.all(path, composedMiddleware);
  }

  sendTo(room: string, data: ArrayBufferLike | string) {
    return this.adapter?.publish(room, data);
  }

  sendJsonTo(room: string, data: any) {
    return this.adapter?.publish(room, JSON.stringify(data));
  }

  private notFound(socket: Socket) {
    socket.end('HTTP/1.1 404 Not Found\r\n\r\n');
  }
}

export default (app: Application) => {
  // this is not a multi clients plugin, so it won't use app.addSingleton
  app.ws = new EggWsServer(app.config.websocket, app);
};
